pkg/engine/mutate/strategicPreprocessing_test.go

```diff
@@ -125,6 +125,130 @@ func Test_preProcessStrategicMergePatch_AnnotationMap(t *testing.T) {
 	}
 }
 
+func Test_preProcessStrategicMergePatch_DifferentKinds(t *testing.T) {
+	rawPolicy := []byte(`{
+		"spec": {
+		  "imagePullSecrets": [
+			{
+			  "name": "regcred"
+			}
+		  ],
+		  "initContainers": [
+			{
+			  "(image)": "index.docker.io/*"
+			}
+		  ]
+		}
+	  }`)
+
+	rawResource := []byte(`{
+		"spec": [
+			"containers": {
+				"{\"name\":\"nginx\"}": {
+				  "image": {},
+				  "imagePullPolicy": {},
+				  "name": {},
+				  "resources": {}
+				}
+			}
+		]
+	}`)
+
+	expected := `{"spec": {"imagePullSecrets": [{"name": "regcred"}], "initContainers": [{}]}}`
+
+	preProcessedPolicy, err := preProcessStrategicMergePatch(string(rawPolicy), string(rawResource))
+	assert.NilError(t, err)
+	output, err := preProcessedPolicy.String()
+	assert.NilError(t, err)
+	re := regexp.MustCompile(`\n`)
+	if !assertnew.Equal(t, strings.ReplaceAll(expected, " ", ""), strings.ReplaceAll(re.ReplaceAllString(output, ""), " ", "")) {
+		t.FailNow()
+	}
+}
+
+func Test_preProcessStrategicMergePatch_InitContainersMatchesTheImage(t *testing.T) {
+	rawPolicy := []byte(`{
+		"spec": {
+		  "imagePullSecrets": [
+			{
+			  "name": "regcred"
+			}
+		  ],
+		  "initContainers": [
+			{
+			  "(image)": "index.docker.io/*",
+			  "name": "name"
+			}
+		  ]
+		}
+	  }`)
+
+	rawResource := []byte(`{
+		"spec": {
+			"initContainers": [
+				{
+					"image": "index.docker.io/someimage",
+					"name": "nginx"
+				}
+			]
+		}
+	}`)
+
+	expected := `{"spec": {"imagePullSecrets": [{"name": "regcred"}], "initContainers": [{"name":"name"}]}}`
+
+	preProcessedPolicy, err := preProcessStrategicMergePatch(string(rawPolicy), string(rawResource))
+	assert.NilError(t, err)
+	output, err := preProcessedPolicy.String()
+	assert.NilError(t, err)
+	re := regexp.MustCompile(`\n`)
+	if !assertnew.Equal(t, strings.ReplaceAll(expected, " ", ""), strings.ReplaceAll(re.ReplaceAllString(output, ""), " ", "")) {
+		t.FailNow()
+	}
+}
+
+func Test_preProcessStrategicMergePatch_InitContainersDoesNotMatchTheImage(t *testing.T) {
+	rawPolicy := []byte(`{
+		"spec": {
+		  "imagePullSecrets": [
+			{
+			  "name": "regcred"
+			}
+		  ],
+		  "initContainers": [
+			{
+			  "(image)": "index.docker.io/*",
+			  "port": "1313",
+			  "foo": "bar",
+			  "bar": "mar",
+			  "dir": "dr"
+			}
+		  ]
+		}
+	  }`)
+
+	rawResource := []byte(`{
+		"spec": {
+			"initContainers": [
+				{
+					"image": "someimage",
+					"name": "nginx"
+				}
+			]
+		}
+	}`)
+
+	expected := `{"spec": {"imagePullSecrets": [{"name": "regcred"}], "initContainers": []}}`
```

Comment:
If the image does not match, we should not apply the policy. Did we test these policies end to end?

Reply:
Policy is not applied here. Patches for initContainers are empty.